pkg/lib/oauth/handler/service_token.go

```diff
@@ -39,23 +38,36 @@ func (s *TokenService) IssueOfflineGrant(
 	client *config.OAuthClientConfig,
 	opts IssueOfflineGrantOptions,
 	resp protocol.TokenResponse,
-) (*oauth.OfflineGrant, error) {
+) (offlineGrant *oauth.OfflineGrant, tokenHash string, err error) {
 	token := s.GenerateToken()
+	tokenHash = oauth.HashToken(token)
 	now := s.Clock.NowUTC()
 	accessEvent := access.NewEvent(now, s.RemoteIP, s.UserAgentString)
 
-	offlineGrant := &oauth.OfflineGrant{
-		AppID:           string(s.AppID),
-		ID:              uuid.New(),
+	refreshToken := &oauth.OfflineGrantRefreshToken{
+		TokenHash:       tokenHash,
+		ClientID:        client.ClientID,
+		CreatedAt:       now,
+		Scopes:          opts.Scopes,
 		AuthorizationID: opts.AuthorizationID,
+	}
+
+	offlineGrant = &oauth.OfflineGrant{
+		AppID:        string(s.AppID),
+		ID:           uuid.New(),
+		IDPSessionID: opts.IDPSessionID,
+		IdentityID:   opts.IdentityID,
+
+		// These fields exist in OfflineGrantRefreshToken too
+		// They need to be provided to handled cases that we refer to an offline grant without token hash
+		// Such as using an id_token
 		ClientID:        client.ClientID,
-		IDPSessionID:    opts.IDPSessionID,
-		IdentityID:      opts.IdentityID,
+		Scopes:          opts.Scopes,
+		AuthorizationID: opts.AuthorizationID,
+		TokenHash:       tokenHash,
```

Comment:
I decided to keep these fields in the top level too, because there are cases that referencing an offline grant without an actual token. (The sid in id_token)

See if any thoughts or concern.